presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergUpdateablePageSource.java

```diff
@@ -25,76 +25,125 @@
+        this.updateRowIdChildColumnIndexes = updateRowIdColumn
+                .map(column -> new int[column.getColumnIdentity().getChildren().size()])
+                .orElse(new int[0]);
+        Map<ColumnIdentity, Integer> columnToIndex = IntStream.range(0, delegateColumns.size())
+                .boxed()
+                .collect(toImmutableMap(idx -> delegateColumns.get(idx).getColumnIdentity(), identity()));
+        updateRowIdColumn.ifPresent(column -> {
+            List<ColumnIdentity> rowIdFields = column.getColumnIdentity().getChildren();
+            for (int i = 0; i < rowIdFields.size(); i++) {
+                ColumnIdentity columnIdentity = rowIdFields.get(i);
+                updateRowIdChildColumnIndexes[i] = requireNonNull(columnToIndex.get(columnIdentity), () -> format("Column %s not found in requiredColumns", columnIdentity));
+                columnIdToRowIdColumnIndex.put(columnIdentity, i);
+            }
+        });
 
-        prefilledBlocks = new Block[size];
-        delegateIndexes = new int[size];
-
-        int outputIndex = 0;
-        int delegateIndex = 0;
-        for (IcebergColumnHandle column : columns) {
-            if (partitionKeys.containsKey(column.getId())) {
-                HivePartitionKey icebergPartition = partitionKeys.get(column.getId());
-                Type type = column.getType();
-                Object prefilledValue = deserializePartitionValue(type, icebergPartition.getValue().orElse(null), column.getName());
-                prefilledBlocks[outputIndex] = nativeValueToBlock(type, prefilledValue);
-                delegateIndexes[outputIndex] = -1;
+        if (!updatedColumns.isEmpty()) {
+            for (int columnIndex = 0; columnIndex < updatedColumns.size(); columnIndex++) {
+                IcebergColumnHandle updatedColumn = updatedColumns.get(columnIndex);
+                columnIdentityToUpdatedColumnIndex.put(updatedColumn.getColumnIdentity(), columnIndex);
             }
-            else if (column.getColumnType() == PARTITION_KEY) {
-                // Partition key with no value. This can happen after partition evolution
-                Type type = column.getType();
-                prefilledBlocks[outputIndex] = nativeValueToBlock(type, null);
-                delegateIndexes[outputIndex] = -1;
+        }
+        for (int i = 0; i < outputColumnToDelegateMapping.length; i++) {
+            if (outputColumns.get(i).isUpdateRowIdColumn()) {
+                continue;
             }
-            else if (IcebergMetadataColumn.isMetadataColumnId(column.getId())) {
-                prefilledBlocks[outputIndex] = nativeValueToBlock(column.getType(), metadataValues.get(column.getColumnIdentity().getId()));
-                delegateIndexes[outputIndex] = -1;
+
+            if (!columnToIndex.containsKey(outputColumns.get(i).getColumnIdentity())) {
+                // nit: Is it appropriate to throw a PrestoException with error code ICEBERG_MISSING_COLUMN here?
```

Comment:
Was this comment intentionally left here?

Reading through this, my thinking is this is something that could happen and it would be an Iceberg error.  I think it means there's some sort of mismatch between the table schema in Iceberg and the output columns generated by Presto.  So perhaps just remove the comment?

Reply:
Ah it was from a review comment. I think I accidentally pasted the review comment as an actual comment  😅  I will remove it